### DISCUSSION TITLE
Has anyone managed to build Qt with dynamic libraries on macos and linux?

### DISCUSSION THREAD

**Discussion by trileyinvicro** [microsoft/vcpkg#31635] [Q&A] [answered]
Building Qt libraries with static linking works perfectly, but I need to build them as dynamic libraries (this works fine for me on Windows).

Specifically, when I build qt5-imageformats on Linux, I get a linker error where libopenjpeg.a is not added to one of the linker commands (on release only, strangely no issue building debug libs) , and with qt5-base on macos I get an error near the end of the qt5-declarative build:

`15:00:18 CMake Error at scripts/cmake/vcpkg_replace_string.cmake:2 (file):
15:00:18   file failed to open for reading (No such file or directory):
15:00:18 
15:00:18     /Users/jenkins/software/vcpkg/packages/qt5-declarative_x64-osx/include/qt5/QtQml/5.15.2/QtQml/private/qqmljsparser_p.h`

Due to the complexity of the Qt build, I haven't been able to get to the bottom of either of these issues.

**Comment by Osyotr** [comment, marked as answer]
Qt5 builds fine with `x64-linux-dynamic`. Maybe update vcpkg? 5.15.2 is pretty old, current version is 5.15.9. Regarding qt5-imageformats issue, it's hard to tell something without full logs.

**Reply by trileyinvicro** [reply]
Thanks, I'll try it with that triplet on Linux. I didn't account for the RPATH before, but I am not certain if it helps my issue.

Does it have to be all dynamic for all of the libraries though? My triplet looks like

`
set(VCPKG_TARGET_ARCHITECTURE x64)
set(VCPKG_CRT_LINKAGE dynamic)
set(VCPKG_LIBRARY_LINKAGE static)

set(VCPKG_CMAKE_SYSTEM_NAME Linux)

if(PORT MATCHES "qt5-")
    set(VCPKG_LIBRARY_LINKAGE dynamic)
endif()
`

**Reply by Osyotr** [reply]
> Does it have to be all dynamic for all of the libraries though?

Should be fine even if only qt5 is dynamic.

**Reply by trileyinvicro** [reply]
Okay thanks. I got the Linux build to work at least with this manifest and just a simple `vcpkg install --overlay-triplets=. --triplet=my-linux-triplet`

`{
    "name": "example",
    "version-date": "2023-03-21",
    "builtin-baseline": "582a7ef7a0dc5cdda5c015e98f57c97218e18c32",
    "dependencies": [
        "qt5-tools",
        {
            "name": "qt5-base",
            "default-features": false
        }
    ]
}`

I hope it's not somehow an interaction with the other libraries I'm using because I have quite a few more.

**Reply by Osyotr** [reply]
Note that it's recommended to also set `VCPKG_FIXUP_ELF_RPATH` on linux. https://github.com/microsoft/vcpkg/blob/master/triplets/community/x64-linux-dynamic.cmake#L7

**Reply by trileyinvicro** [reply]
Turns out I needed to update the baseline, now I can build dynamic libs.
